File: problems/solutions/problem1-two-sum.py
"""
问题1：两数之和
使用哈希表解法
时间复杂度：O(n)
空间复杂度：O(n)
"""

def two_sum(nums, target):
    hash_map = {}
    for i, num in enumerate(nums):
        complement = target - num
        if complement in hash_map:
            return [hash_map[complement], i]
        hash_map[num] = i
    return []

# 输入
n, target = map(int, input().split())
nums = list(map(int, input().split()))

# 求解
result = two_sum(nums, target)
print(result[0], result[1])

File: problems/solutions/problem1-two-sum.cpp
/*
 * 问题1：两数之和
 * 使用哈希表解法
 * 时间复杂度：O(n)
 * 空间复杂度：O(n)
 */

#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

int main() {
    int n, target;
    cin >> n >> target;
    
    vector<int> nums(n);
    for (int i = 0; i < n; i++) {
        cin >> nums[i];
    }
    
    unordered_map<int, int> hash;
    
    for (int i = 0; i < n; i++) {
        int complement = target - nums[i];
        if (hash.count(complement)) {
            cout << hash[complement] << " " << i << endl;
            return 0;
        }
        hash[nums[i]] = i;
    }
    
    return 0;
}
